### PR TITLE
Add error types to policy

### PR DESCRIPTION
**Pull request by dstiliadis**
Adds additional error types in the policy object.

### CODE REVIEW

**Comment by ghost** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "engine-refactor",
    "component": "enforcerd",
    "pr-id": "1094",
    "commit-sha": "b211631648f95c68f476566a00b09d2ef14d5a3c"
  },
  {
    "project": "engine-refactor",
    "component": "integrations-lib",
    "pr-id": "95",
    "commit-sha": "2ca15f70ded1f3477eb0f1da4b774bb5c25810dd"
  },
  {
    "project": "engine-refactor",
    "component": "trireme-lib",
    "pr-id": "765",
    "commit-sha": "e33befde874d0ede2895dac5c1a32fdba2ea030d"
  }
]
```

**Comment by ghost** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "engine-refactor",
    "component": "integrations-lib",
    "pr-id": "95",
    "commit-sha": "2ca15f70ded1f3477eb0f1da4b774bb5c25810dd"
  },
  {
    "project": "engine-refactor",
    "component": "trireme-lib",
    "pr-id": "765",
    "commit-sha": "e33befde874d0ede2895dac5c1a32fdba2ea030d"
  },
  {
    "project": "engine-refactor",
    "component": "enforcerd",
    "pr-id": "1094",
    "commit-sha": "b211631648f95c68f476566a00b09d2ef14d5a3c"
  },
  {
    "project": "engine-refactor",
    "component": "apotests",
    "pr-id": "1991",
    "commit-sha": "925e9ef469c65bf5b13add696830e92945922f7f"
  }
]
```

**Comment by ghost** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "engine-refactor",
    "component": "apotests",
    "pr-id": "1991",
    "commit-sha": "925e9ef469c65bf5b13add696830e92945922f7f"
  },
  {
    "project": "engine-refactor",
    "component": "integrations-lib",
    "pr-id": "95",
    "commit-sha": "2ca15f70ded1f3477eb0f1da4b774bb5c25810dd"
  },
  {
    "project": "engine-refactor",
    "component": "trireme-lib",
    "pr-id": "765",
    "commit-sha": "e33befde874d0ede2895dac5c1a32fdba2ea030d"
  },
  {
    "project": "engine-refactor",
    "component": "enforcerd",
    "pr-id": "1094",
    "commit-sha": "2907a31c00dd69c996f14e98f39349aa0fa82799"
  }
]
```

**Comment by ghost** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "engine-refactor",
    "component": "integrations-lib",
    "pr-id": "95",
    "commit-sha": "2ca15f70ded1f3477eb0f1da4b774bb5c25810dd"
  },
  {
    "project": "engine-refactor",
    "component": "trireme-lib",
    "pr-id": "765",
    "commit-sha": "e33befde874d0ede2895dac5c1a32fdba2ea030d"
  },
  {
    "project": "engine-refactor",
    "component": "enforcerd",
    "pr-id": "1094",
    "commit-sha": "2907a31c00dd69c996f14e98f39349aa0fa82799"
  },
  {
    "project": "engine-refactor",
    "component": "apotests",
    "pr-id": "1991",
    "commit-sha": "925e9ef469c65bf5b13add696830e92945922f7f"
  }
]
```

**Comment by ghost** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "engine-refactor",
    "component": "trireme-lib",
    "pr-id": "765",
    "commit-sha": "e33befde874d0ede2895dac5c1a32fdba2ea030d"
  },
  {
    "project": "engine-refactor",
    "component": "enforcerd",
    "pr-id": "1094",
    "commit-sha": "2907a31c00dd69c996f14e98f39349aa0fa82799"
  },
  {
    "project": "engine-refactor",
    "component": "apotests",
    "pr-id": "1991",
    "commit-sha": "3ba9dba8ea9f6e84d679e806b6e9dbb1f5c540e5"
  },
  {
    "project": "engine-refactor",
    "component": "integrations-lib",
    "pr-id": "95",
    "commit-sha": "2ca15f70ded1f3477eb0f1da4b774bb5c25810dd"
  }
]
```

File: controller/internal/enforcer/applicationproxy/applicationproxy.go [hunk 1]
```diff
@@ -103,17 +103,17 @@ func (p *AppProxy) Enforce(ctx context.Context, puID string, puInfo *policy.PUIn
 
 	data, err := p.puFromID.Get(puID)
 	if err != nil || data == nil {
-		return fmt.Errorf("Undefined PU - Context not found: %s", puID)
+		return fmt.Errorf("undefined PU - Context not found: %s", puID)
```
Comment (primalmotion) [inline controller/internal/enforcer/applicationproxy/applicationproxy.go:106]: no dash

Comment (primalmotion) [reply controller/internal/enforcer/applicationproxy/applicationproxy.go:106]: lower case Context

File: controller/internal/enforcer/applicationproxy/applicationproxy.go [hunk 2]
```diff
@@ -123,7 +123,7 @@ func (p *AppProxy) Enforce(ctx context.Context, puID string, puInfo *policy.PUIn
 	if c, cerr := p.clients.Get(puID); cerr == nil {
 		_, perr := p.processCertificateUpdates(puInfo, c.(*clientData), caPool)
 		if perr != nil {
-			zap.L().Error("Failed to update certificates and services", zap.Error(perr))
+			zap.L().Error("unable to update certificates and services", zap.Error(perr))
```
Comment (primalmotion) [inline controller/internal/enforcer/applicationproxy/applicationproxy.go:126]: zap log is Capitalized :)

File: controller/internal/enforcer/enforcer.go [hunk 3]
```diff
@@ -94,19 +94,19 @@ func (e *enforcer) Enforce(contextID string, puInfo *policy.PUInfo) error {
 
 	if e.transport != nil {
 		if err := e.transport.Enforce(contextID, puInfo); err != nil {
-			return fmt.Errorf("Failed to enforce in nfq: %s", err)
+			return fmt.Errorf("unable to enforce in nfq: %s", err)
```
Comment (primalmotion) [inline controller/internal/enforcer/enforcer.go:97]: can you provide more info like the pu ID etc?

File: controller/internal/enforcer/enforcer.go [hunk 4]
```diff
@@ -94,19 +94,19 @@ func (e *enforcer) Enforce(contextID string, puInfo *policy.PUInfo) error {
...
 		}
 	}
 
 	if e.proxy != nil {
 		if err := e.proxy.Enforce(context.Background(), contextID, puInfo); err != nil {
-			return fmt.Errorf("Failed to enforce in proxy: %s", err)
+			return fmt.Errorf("unable to enforce in proxy: %s", err)
```
Comment (primalmotion) [inline controller/internal/enforcer/enforcer.go:103]: same

File: controller/pkg/usertokens/oidc/oidc.go [hunk 5]
```diff
@@ -68,9 +67,9 @@ func NewClient(ctx context.Context, v *TokenVerifier) (*TokenVerifier, error) {
 	// If it is not a compliant provider we should report and error here.
 	provider, err := oidc.NewProvider(ctx, v.ProviderURL)
 	if err != nil {
-		zap.L().Error("Failed to initialize OIDC provider", zap.Error(err), zap.String("Provider URL", v.ProviderURL))
 		return nil, fmt.Errorf("Failed to initialize provider: %s", err)
```
Comment (primalmotion) [inline controller/pkg/usertokens/oidc/oidc.go:70]: the url of the provider would help

**Review by primalmotion — COMMENTED** [review]
error needs to give a bit more context

**Pull request merged**